### PR TITLE
askrene: fixed a timeout corner case

### PR DESCRIPTION
**Pull request by Lagrang3**
We've introduced a time out protection in #8636 to avoid hitting infinite loops
in askrene.

I've found a test case that hits the time out, which we don't usually want. We should be smart
enough to detect deadlocks and give up with something like "Payment is unfeasible".

To reproduce the condition we need two channels the first with an HTLC max = x
and the second with a base fee = x. So that any amount f to forward to the 3rd
node requires at least x+f to pass through the first channel, which is not
possible due to the HTLC max limit.
Refine removes all sats from the route and linear routes' while loop continues
coming back to the exact same situation at each iteration. We only end this
timing out.

We fix the infinite loop by deciding (under some arbitrary condition)
when a path has a bottleneck and then disabling one channel in that path that causes the bottleneck.

> [!IMPORTANT]
>
> 26.04 FREEZE March 11th: Non-bugfix PRs not ready by this date will wait for 26.06.
>
> RC1 is scheduled on _March 23rd_
>
> The final release is scheduled for April 15th.

## Checklist
Before submitting the PR, ensure the following tasks are completed. If an item is not applicable to your PR, please mark it as checked:

- [ ] The changelog has been updated in the relevant commit(s) according to the [guidelines](https://docs.corelightning.org/docs/coding-style-guidelines#changelog-entries-in-commit-messages).
- [ ] Tests have been added or modified to reflect the changes.
- [ ] Documentation has been reviewed and updated as needed.
- [ ] Related issues have been listed and linked, including any that this PR closes.
- [ ] *Important* All PRs must consider how to reverse any persistent changes for `tools/lightning-downgrade`

### CODE REVIEW

File: plugins/askrene/refine.c [hunk 1]
```diff
@@ -494,10 +533,43 @@ const char *refine_flows(const tal_t *ctx, struct route_query *rq,
...
+		}
+
+		/* A path with a very small deliverable amount is not worth the
+		 * effort, and we don't want either a path that for fees and
+		 * HTLC max constraints removes too much from the actual
+		 * delivery amount. In theory the MCF already has partitioned
+		 * the payment in different paths. The refinement step is not
+		 * expected to change the flow by much. */
+		if (amount_msat_less(deliverable, AMOUNT_MSAT(1000000)) ||
+		    amount_msat_ratio(deliverable, try_deliver) < 0.5) {
```
Comment (Lagrang3) [inline plugins/askrene/refine.c]: In the previous version we simply made `flow->delivers = min(flow->delivers, max_deliverable(flow->path))`. This was a simple idea to reduce the flow amount in order to avoid hitting HTLC max constraints after adding fees. But we ended up in `flow->delivers = 0` in some rare cases, maybe even not happening in real life at least not by chance (I can't discard the possibility that someone could try to disrupt the network on purpose), and this cases lead to an endless loop in `linear_routes`.

To avoid repeating the same situation again and again we need to change something.
My idea is to detect those bad paths and disable a channel there, this is done below with function `remove_bottleneck`, so that we don't use them again for this payment: infinite loop is adverted.

The question on what is our definition of a *bad path* is what makes this `if` condition arbitrary.
The baddest of all are those that can't deliver nothing, ie. `deliverable == 0`. I relaxed the condition a bit by including less bad paths as well, ie. `deliverable<1000sat` or `deliverable / flow->delivers < 0.5`. I am open to suggestions here.

File: plugins/askrene/explain_failure.c [hunk 2]
```diff
@@ -290,7 +290,7 @@ const char *explain_failure(const tal_t *ctx,
 			      explanation);
 	}
 
-	return rq_log(ctx, rq, LOG_BROKEN,
+	return rq_log(ctx, rq, LOG_UNUSUAL,
```
Comment (rustyrussell) [inline plugins/askrene/explain_failure.c]: Well, if we're going to do this... which we probably shouldn't (I would rather remember we hit this weird case and hand a flag into explain_failure), we should definitely use the CI_UNEXPECTED constant, which will trigger CI failure if this happens...

**Review by rustyrussell — CHANGES_REQUESTED** [review]
I know it annoying, but I think we have to plumb one "bottleneck_scid" all the way back to explain_failure.

0x0x0 == we didn't use the bottleneck logic.
anything else == this is one scid we discarded because its configuration was restricting the flow.

Then expain_failure can use that if it can't find another reason, and it's on the "obvious" path ("1x2x3 is configured with very high fees or low capacity").

File: plugins/askrene/explain_failure.c (hunk 2 above)
Comment (Lagrang3) [reply plugins/askrene/explain_failure.c]: Alright. We will pull the bottleneck scid to explain the failure and then if that still doesn't explain it then we LOG_BROKEN.

Comment (Lagrang3) [reply plugins/askrene/explain_failure.c]: Same logic should apply if we disable a channel for an HTLC min violation during refine, see `remove_htlc_min_violations`.

**Comment by Lagrang3** [comment]
> I know it annoying, but I think we have to plumb one "bottleneck_scid" all the way back to explain_failure.
> 
> 0x0x0 == we didn't use the bottleneck logic. anything else == this is one scid we discarded because its configuration was restricting the flow.
> 
> Then expain_failure can use that if it can't find another reason, and it's on the "obvious" path ("1x2x3 is configured with very high fees or low capacity").

I have a second thought on this after reading `explain_failure`. This kind of bottleneck can trigger one of the possible explanations branches in `explain_failure`. It doesn't right now because we don't add fees along the way. I am going to fix it.

File: plugins/askrene/explain_failure.c (hunk 2 above)
Comment (Lagrang3) [reply plugins/askrene/explain_failure.c]: I left `LOG_BROKEN` here. I see the logic. We should be able to deduce why we couldn't find a route.
But in my opinion this looks more than an *unusual* situation than a signal that our plugin is *broken*.
To be honest I don't expect `explain_failure` to find every reason all of the time. The reasons can
be very complex once you need to start splitting the payment into parts.

**Comment by Lagrang3** [comment]
> I know it annoying, but I think we have to plumb one "bottleneck_scid" all the way back to explain_failure.
> 
> 0x0x0 == we didn't use the bottleneck logic. anything else == this is one scid we discarded because its configuration was restricting the flow.
> 
> Then expain_failure can use that if it can't find another reason, and it's on the "obvious" path ("1x2x3 is configured with very high fees or low capacity").

Fixed it.
There was no need to pass the scid to the caller. The bottleneck is seen by `explain_failure` once you make it compute
the amounts along the route taking fees into consideration, which was missing.

File: tests/test_askrene.py [hunk 3]
```diff
@@ -2433,7 +2433,6 @@ def test_includefees(node_factory):
     )
 
 
-@unittest.skip
```
Comment (rustyrussell) [inline tests/test_askrene.py:2436]: We usually use an explicit "@pytest.mark.xfail(strict=True)" here BTW.  This still runs the test, but REQUIRES that it fail.

Comment (Lagrang3) [reply tests/test_askrene.py:2436]: Oh, thanks. I'll keep that in mind for the next time.
